### PR TITLE
feat: add Claude Agent (A2A)

### PR DESCRIPTION
**Pull request by huang195**
## Summary

Adds a new A2A agent, **`claude_agent`**, that drives the `claude` CLI headlessly. Each A2A `context_id` maps to a persistent Claude session with its own isolated working directory, so the agent serves concurrent sessions from one user and sessions from multiple users. The model is reached through a LiteLLM (Anthropic-compatible) endpoint.

### What's included
- `a2a/claude_agent/` — the agent: `configuration`, `session` (uuid5-derived session + per-session ephemeral workdir, per-session lock, LRU registry with off-loop cleanup), `runner` (one `claude -p` subprocess per turn; `--session-id` then `--resume`; `--dangerously-skip-permissions`; concurrent stdout/stderr drain; timeout; kills the subprocess on cancel/error), `events` (stream-json → A2A rich progress + final artifact), `agent` (A2A wiring).
- Pinned `a2a-sdk[http-server]==0.3.26` (pydantic API, matching the other agents).
- `Dockerfile` (Node + the `claude` CLI + uv + Python, non-root; pinned `claude-code@2.1.161` + `uv 0.11.18`), `.env.litellm`, README with a trust-model section and GHCR-based deploy instructions.
- `sample-environments.yaml`: a `claude-agent` env profile.
- `build.yaml`: `claude_agent` added to the publish matrix so it ships to `ghcr.io/kagenti/agent-examples/claude_agent`.

Config: required values are `ANTHROPIC_AUTH_TOKEN` and `ANTHROPIC_BASE_URL` (your LiteLLM endpoint); `ANTHROPIC_MODEL` defaults to `sonnet`.

Deployment manifests live in `kagenti/kagenti` (`examples/agents/`).

> Naming: the agent is called **claude-agent** (not "claude-code-agent"). It still wraps Anthropic's `@anthropic-ai/claude-code` CLI internally — that package name and the `CLAUDE_CODE_*` env vars it reads are kept as the only references to the product name.

## Test Plan
- [x] `uv run pytest -v` — 24 unit tests pass (configuration, session isolation/LRU, event translation, runner happy/timeout/non-zero-exit/cancel paths, agent executor + cancel).
- [x] `ruff check` + `ruff format --check` clean (CI ruff 0.11.4).
- [x] `docker build` + container smoke test — serves the agent card, `claude` CLI on PATH.
- [x] Verified end-to-end in the Kagenti UI (Kind): created via Deploy-from-Existing-Image, chatted with the agent.

Assisted-By: Claude Code

### CODE REVIEW

File: a2a/claude_agent/src/claude_agent/agent.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    app = server.build()
+
+    # Serve the newer agent-card.json path alongside the legacy agent.json path.
+    app.routes.insert(
+        0,
+        Route(
+            "/.well-known/agent-card.json",
+            server._handle_get_agent_card,
+            methods=["GET"],
+            name="agent_card_new",
```
Comment (cwiklik) [inline a2a/claude_agent/src/claude_agent/agent.py]: **suggestion**: `server._handle_get_agent_card` accesses a private (`_`-prefixed) method of `A2AStarletteApplication`. This couples to an internal API that will break silently if the SDK renames or removes it. Consider implementing a standalone handler that returns `agent_card.model_dump_json()` with the right content-type, or opening a feature request on the a2a-sdk to expose this publicly.

File: a2a/claude_agent/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Claude Code needs Node (the CLI is an npm package); the A2A server is Python.
+FROM node:22-bookworm-slim
```
Comment (cwiklik) [inline a2a/claude_agent/Dockerfile]: **nit**: `node:22-bookworm-slim` is a rolling tag. For supply-chain reproducibility, consider pinning to a SHA digest (`node:22-bookworm-slim@sha256:...`). Fine for an example agent, but production deploys would want the pin.

**Review by cwiklik — APPROVED** [review]
Clean, well-structured implementation. The trust model for `--dangerously-skip-permissions` is documented thoroughly — container/pod is the isolation boundary, prompts are fully trusted. Subprocess management has proper cleanup in the `finally` block (no orphaned processes). Session registry's LRU eviction correctly skips locked sessions and offloads `rmtree` off the event loop. 19 tests cover all modules including edge cases (timeout, non-zero exit, cancel, eviction).

**Areas reviewed**: Python, Dockerfile, CI (build matrix), tests, security
**Commits**: 15/15 signed-off
**CI**: all 10 checks passing

File: a2a/claude_agent/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+## Test
+
+```bash
+uv run pytest -v
+```
+
+## Deploy on Kagenti
+
+The image is published to GHCR by the repo's `Build-Publish` workflow (on `v*`
```
Comment (maia-iyer) [inline a2a/claude_agent/README.md:50]: Wait we shouldn't be publishing this to comply with Claude Code's terms of use as far as I understand

**Review by maia-iyer — CHANGES_REQUESTED** [review]

**Review by evaline-ju — COMMENTED** [review]
I noticed while trying this out that every conversation turn created the "Claude Code session started" event 

Image

This doesn't have to be fixed in this PR but should likely be noted as a design gap - with each turn spawning a new process and the session files living in jsonl by cwd+uuid, each `--resume` turn 2 onwards reloads conversation history before processing each message. if after a container restart the particular file can't be found, a fresh convo would be started but appear to continue from the user perspective because the same a2a context ID is used.

File: a2a/claude_agent/README.md (hunk 3 above)
Comment (maia-iyer) [reply a2a/claude_agent/README.md:50]: after offline conversation - seems we cannot find a clause in the terms of use that specify so long as we are not trafficking through our own solutions or providing our own credentials for public use

**Review by pdettori — COMMENTED** [review]
## Suggestion: Consider using the Claude Agent SDK instead of the CLI subprocess

Nice work on the concurrency model and A2A wiring — this is well-engineered. However, I want to flag that Anthropic now offers a **[Claude Agent SDK](https://code.claude.com/docs/en/agent-sdk/overview)** (Python and TypeScript) that provides the same capabilities as a library rather than requiring a subprocess wrapper around the CLI.

The core runner (~140 lines of subprocess management) would collapse to roughly:

```python
from claude_agent_sdk import query, ClaudeAgentOptions

async for message in query(
    prompt=user_prompt,
    options=ClaudeAgentOptions(
        resume=session_uuid if session.started else None,
        allowed_tools=["Read", "Edit", "Bash", "Glob", "Grep"],
        model=config.model,
    ),
):
    yield translate_to_a2a(message)
```

### Advantages of the Agent SDK over the CLI approach

| Area | CLI (current) | Agent SDK |
|------|---------------|----------|
| **Container image** | Needs `node:22-bookworm-slim` base (Node.js required for npm-installed CLI) → larger image, more CVE surface | Pure Python image — no Node.js dependency at all |
| **Security model** | `--dangerously-skip-permissions` (all-or-nothing) — any prompt can execute arbitrary code | `allowed_tools=[...]` — granular per-tool permission control |
| **Performance** | Process-per-turn: spawns Node.js, loads Claude runtime, rehydrates session from disk on every turn | In-process: state stays in memory across turns, no cold-start overhead |
| **Stream format stability** | `--output-format stream-json` is not a documented stable API — can change between CLI versions | Stable Python API contract with semantic versioning |
| **Complexity** | Requires: subprocess lifecycle management, concurrent stdout/stderr drain, process-kill cleanup, stream-json parser (`events.py`) | All handled internally by the SDK — you iterate async messages |
| **Session management** | Transcripts on local disk (`~/.claude/projects/...`); `--session-id` / `--resume` flags | Same disk-based JSONL, but `resume=session_id` is a first-class option; easier to back with PVC or external storage |
| **Maintenance** | Pinned `@anthropic-ai/claude-code@2.1.161` — version coupling to undocumented internals | `pip install claude-agent-sdk` with standard semver guarantees |
| **Hooks** | Not available (skip-permissions is the only mode) | `PreToolUse` / `PostToolUse` callbacks for audit logging, validation, blocking |

### Key question

Does the Python Agent SDK support routing through a LiteLLM-compatible base URL (your `ANTHROPIC_BASE_URL` pattern)? The SDK supports `ANTHROPIC_API_KEY` natively and has Bedrock/Vertex/Azure auth built in. If LiteLLM routing works (or can be made to work via the SDK's API key + base URL config), there's no architectural reason to keep the CLI subprocess wrapper.

### Recommendation

This PR works and is well-tested — but I'd suggest either:
1. **Refactor now** to use `claude-agent-sdk` before the CLI-based approach establishes inertia, or
2. **Merge as-is** with a tracked follow-up issue for SDK migration, noting that the stream-json format is an internal detail that may break on CLI upgrades.

Either way, great foundation for the A2A integration — the session registry, LRU eviction, and event translation patterns will carry over nicely to an SDK-based implementation.

**Review by maia-iyer — APPROVED** [review]
